### PR TITLE
[new_profile] Fix 6 digit years & month functionality

### PR DESCRIPTION
**Pull request by zaliqarosli**
fixes https://github.com/aces/Loris/issues/4728 and https://github.com/aces/Loris/issues/4564

to test:
- [ ] on test.loris.ca or 21.0-release branch, type in the year by keyboard into any date element. you will be able to enter a 6 digit year
- [ ] on this branch, you can only enter 4 digits on Chrome. The max attribute is broken on Firefox
- [ ] repeat the above with dob format in Configurations set to 'Ym' or 'YM'

caveat: Firefox allows 5/6 digit years because the "max" attribute breaks on that browser

### CODE REVIEW

File: modules/new_profile/jsx/NewProfileIndex.js [hunk 1]
```diff
@@ -194,7 +194,7 @@ class NewProfileIndex extends React.Component {
             name = "dateTakenConfirm"
             label = "Date of Birth Confirm"
             minYear = "2000"
-            maxYear = "2017"
+            maxYear = "2017-12-31"
```
Comment (johnsaigle) [inline modules/new_profile/jsx/NewProfileIndex.js]: Should these values maybe not be hard coded? I'm not sure why there would be a boundary of 2000-2017 specifically for every study. I think this should be pulled from the `ageMin and `ageMax` config settings.

I know that's not really the scope of this PR so if you don't want to fix it just let me know and I'll open an issue for it.

Comment (zaliqarosli) [reply modules/new_profile/jsx/NewProfileIndex.js]: yes, you're definitely right!

Comment (zaliqarosli) [reply modules/new_profile/jsx/NewProfileIndex.js]: and no worries, i'll fix it

**Comment by zaliqarosli** [comment]
@johnsaigle i made some change. check em out! it does however only work on chrome, and not firefox. so i'm gonna have to find the rest of the solution if it exists

**Comment by johnsaigle** [comment]
Working for me on Chrome! (but still not on Firefox)

**Comment by zaliqarosli** [comment]
@johnsaigle i've looked everywhere on the internet and looks like its a browser issue that Firefox doesn't allow the max attribute to work properly. it looks like there's nothing we can do about it.

@ridz1208 can you bless the fact that any project with a 'YM' dob format using Firefox will have to fill the date in via a textbox and not a date picker?

**Comment by johnsaigle** [comment]
There is actually a date picker calendar thing! It should doesn't respond the same way when you type the numbers manually. I don't think it's a big issue.

**Comment by zaliqarosli** [comment]
@johnsaigle there is a date picker for input type="date", but now input type="month" on firefox

**Comment by johnsaigle** [comment]
Ah okay, I misunderstood.

**Comment by zaliqarosli** [comment]
@johnsaigle no worries! minus the caveats that I just added to the description, is everything else working fine in this PR? please approve and pass manual test if all good! thanks!

**Comment by johnsaigle** [comment]
>any project with a 'YM' dob format using Firefox will have to fill the date in via a textbox and not a date picker

@zaliqarosli @kongtiaowang It might be good to update the `new_profile` test plan to include this. Could be done in PR #4686.

**Comment by ridz1208** [comment]
@zaliqarosli to be clear, on LORIS v20.3.0, firefox is rendering a year month field (not text box) because of the polyfills we have. The polyfills currently dont work on LORIS 21.0.0 is that correct ?

is the backend validation still gonna catch the error in formatting of the date ? do we give a hint of the formatting of the date that needs to be entered to the user ?

Unless i'm mistaken, we have a loss of functionality here which might need to be fixed before release. I'm not opposed however to fixing in in another PR if this one resolves half of the problem

**Comment by zaliqarosli** [comment]
okay i'll make a PR to have a date picker for input type month in Firefox for this release! @ridz1208

**Comment by zaliqarosli** [comment]
@ridz1208 the fix ended up being simpler than i thought! i just moved the polyfill to the DateElement and it works exactly as before!

@johnsaigle @kongtiaowang the test plan doesn't need to be updated anymore. and everything should work minus the fact that firefox allows 5/6 digit years. this weird browser bug is still true

**Comment by kongtiaowang** [comment]
@zaliqarosli YMd works well, but YM can't create a new profile.

**Comment by zaliqarosli** [comment]
@kongtiaowang what's the error? is it a front-end issue or a data submission issue?

**Comment by kongtiaowang** [comment]
@zaliqarosli backend error. the YM should match the date type (yyyy-mm-dd) in the SQL table.
Could not execute INSERT INTO candidate SET `RegistrationCenterID`=:RegistrationCenterID,`PSCID`=:PSCID,`DoB`=:DoB,`EDC`=:EDC,`Sex`=:Sex,`ProjectID` = NULL,`Date_active`=:Date_active,`Date_registered`=:Date_registered,`Entity_type`=:Entity_type,`RegisteredBy`=:RegisteredBy,`UserID`=:UserID,`ExternalID`=:ExternalID,`CandID`=:CandID. Stack trace#0 /var/www/loris/php/libraries/Database.class.inc(263): Database->_realinsert('candidate', Array, true)\n#1 /var/www/loris/php/libraries/Candidate.class.inc(278): Database->insert('candidate', Array)\n#2 /var/www/loris/modules/new_profile/php/profile.class.inc(73): Candidate::createNew(1, '2008-01', '2008-01', 'female', 'DCC1014')\n#3 /var/www/loris/src/Middleware/UserPageDecorationMiddleware.php(171): LORIS\\new_profile\\Profile->handle(Object(Zend\\Diactoros\\ServerRequest))\n#4 /var/www/loris/src/Middleware/PageDecorationMiddleware.php(49): LORIS\\Middleware\\UserPageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\new_profile\\Profile))\n#5 /var/www/loris/php/libraries/NDB_Page.class.inc(653): LORIS\\Middleware\\PageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\new_profile\\Profile))\n#6 /var/www/loris/php/libraries/Module.class.inc(341): NDB_Page->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\new_profile\\Profile))\n#7 /var/www/loris/src/Middleware/ResponseGenerator.php(49): Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#8 /var/www/loris/src/Middleware/AuthMiddleware.php(62): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\new_profile\\Module))\n#9 /var/www/loris/src/Router/ModuleRouter.php(97): LORIS\\Middleware\\AuthMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\new_profile\\Module))\n#10 /var/www/loris/src/Router/BaseRouter.php(99): LORIS\\Router\\ModuleRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#11 /var/www/loris/src/Middleware/ResponseGenerator.php(49): LORIS\\Router\\BaseRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#12 /var/www/loris/src/Middleware/ContentLength.php(51): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#13 /var/www/loris/htdocs/index.php(47): LORIS\\Middleware\\ContentLength->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#14 {main}, referer: https://wangshen-dev.loris.ca/new_profile/
[Tue Jun 04 15:15:52.234737 2019] [php7:error] [pid 12955] [client 132.216.56.161:49646] PHP Fatal error:  Maximum execution time of 30 seconds exceeded in /var/www/loris/php/libraries/Database.class.inc on line 704, referer: https://wangshen-dev.loris.ca/new_profile/

**Comment by zaliqarosli** [comment]
@kongtiaowang thanks for testing and reporting!

**Comment by zaliqarosli** [comment]
rebasing this PR to https://github.com/aces/Loris/pull/4834 so that the above reported date validation error is fixed

**Comment by zaliqarosli** [comment]
@cmadjar yes! thanks for reporting this haha the same thing happened to me and i didn't know what happened. i will investigate! thanks

**Comment by cmadjar** [comment]
@zaliqarosli this is failing Travis.

**Comment by zaliqarosli** [comment]
> When the format in the Config module is YM, the date field to enter change to MY (month having to be entered as February instead of 02 for example). For consistency, it would be good that it remains the YM type of entry (Y=4 digits, M=2 digits).

@cmadjar the way the date elements are rendered is the browser's decision and i unfortunately can't change that. in firefox, it renders as YYYY-MM but in chrome, its MMMMMM(words)-YYYY

> Also, when I hit create candidate for the case where the date format is YM, my apache log goes crazy and my sandbox gets frozen until I restart apache. It tries indefinitely to insert into the candidate table with failure

this should be fixed now!

**Comment by cmadjar** [comment]
Retested it with the latest changes and all is good.

File: jsx/Form.js [hunk 2]
```diff
@@ -976,6 +976,27 @@ class DateElement extends Component {
...
   }
 
+  componentDidMount() {
+    if (!Modernizr.inputtypes.month) {
+      let monthInputs = $('input[type=month]');
+      monthInputs.datepicker({
+        dateFormat: 'yy-mm',
+        changeMonth: true,
+        changeYear: true,
+        yearRange: '1900:' + new Date().getFullYear(),
```
Comment (driusan) [inline jsx/Form.js]: unless I'm misunderstanding, isn't this going to limit the year to the current year? I don't think this happens with things not using the polyfill. What if you need to enter a date in the future?

**Comment by zaliqarosli** [comment]
blocked, working on an issue on firefox

**Comment by zaliqarosli** [comment]
@cmadjar @driusan should be fully fixed now to work on both chrome and firefox. please let me know if i'm still missing something. thanks!

**Comment by cmadjar** [comment]
Tested again with the latest changes on Firefox and Chrome. Everything seems to work as expected!

**Pull request merged**